### PR TITLE
feat: Add the ability to create dashboard apps from the UI

### LINKED ISSUE #4934: Dashboard apps UI
UI for creating, updating, and deleting dashboard apps.

### PR DESCRIPTION
**Pull request by fayazara**
## Description

Allows users to create dashboard apps from UI

Demo - https://www.loom.com/share/3020fcd171c04c64be0743710dde3cc5

Fixes #4934

## Type of change

Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Test cases for respective modules

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/javascript/dashboard/i18n/locale/en/integrations.json [hunk 1]
```diff
@@ -81,6 +84,51 @@
     },
     "CONNECT": {
       "BUTTON_TEXT": "Connect"
+    },
+    "DASHBOARD_APPS": {
+      "HEADER_BTN_TXT": "Add new dashboard app",
```
Comment (pranavrajs) [inline app/javascript/dashboard/i18n/locale/en/integrations.json:89]:
```suggestion
      "HEADER_BTN_TXT": "Add a new dashboard app",
```

File: app/javascript/dashboard/routes/dashboard/settings/integrations/Dashboardapps/Index.vue [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+                  size="tiny"
+                  color-scheme="secondary"
+                  class-names="grey-btn"
+                  :is-loading="loading[app.id]"
+                  icon="edit"
+                  @click="editApp(app)"
+                />
+                <woot-button
+                  v-tooltip.top="
+                    $t(
```
Comment (pranavrajs) [inline app/javascript/dashboard/routes/dashboard/settings/integrations/Dashboardapps/Index.vue:59]: Please create a sub component DashboardAppRow

File: app/javascript/dashboard/routes/dashboard/settings/integrations/Index.vue [hunk 3]
```diff
@@ -17,6 +17,16 @@
               :integration-action="item.action"
             />
           </div>
+          <div class="small-12 columns integration">
+            <integration
+              integration-id="dashboard-apps"
+              integration-logo="dashboard-apps.svg"
+              integration-name="Dashboard apps"
+              integration-description="Dashboard apps lets you add custom applications right into chatwoot dashboard, using the conversation context that chatwoot provides. You can read more about it in the documentation."
```
Comment (pranavrajs) [inline app/javascript/dashboard/routes/dashboard/settings/integrations/Index.vue:25]: Use i18n.

File: app/javascript/dashboard/routes/dashboard/settings/integrations/integrations.routes.js [hunk 4]
```diff
@@ -35,6 +36,12 @@ export default {
           name: 'settings_integrations_webhook',
           roles: ['administrator'],
         },
+        {
+          path: 'dashboard-apps',
+          component: Dashboardapps,
```
Comment (pranavrajs) [inline app/javascript/dashboard/routes/dashboard/settings/integrations/integrations.routes.js:41]: DashboardApps

File: app/javascript/dashboard/routes/dashboard/settings/integrations/Dashboardapps/Index.vue [hunk 5]
```diff
@@ -0,0 +1,180 @@
+<template>
```
Comment (pranavrajs) [inline app/javascript/dashboard/routes/dashboard/settings/integrations/Dashboardapps/Index.vue:1]: Rename the folder to DashboardApps

File: app/javascript/dashboard/i18n/locale/en/integrations.json [hunk 6]
```diff
@@ -81,6 +84,51 @@
...
+        "TABLE_HEADER": [
+          "Name",
+          "Type",
+          "URL"
+        ],
+        "EDIT_TOOLTIP": "Edit app",
+        "DELETE_TOOLTIP": "Delete app"
+      },
+      "FORM": {
+        "TITLE_LABEL": "Dashboard app name",
```
Comment (pranavrajs) [inline app/javascript/dashboard/i18n/locale/en/integrations.json:103]:
```suggestion
        "TITLE_LABEL": "App name",
```

The error message says app name so let us be consistent.

File: app/javascript/dashboard/i18n/locale/en/integrations.json [hunk 7]
```diff
@@ -81,6 +84,51 @@
...
+        "TITLE_PLACEHOLDER": "Enter a name for your dashboard app",
+        "TITLE_ERROR": "App name required",
```
Comment (pranavrajs) [inline app/javascript/dashboard/i18n/locale/en/integrations.json:105]:
```suggestion
        "TITLE_ERROR": "A name for the dashboard app is required",
```

File: app/javascript/dashboard/i18n/locale/en/integrations.json (hunk 6 above)
Comment (pranavrajs) [inline app/javascript/dashboard/i18n/locale/en/integrations.json:103]:
```suggestion
        "TITLE_LABEL": "Name",
```

Let us keep it simple.

File: app/javascript/dashboard/i18n/locale/en/integrations.json [hunk 8]
```diff
@@ -81,6 +84,51 @@
...
+        "URL_LABEL": "Iframe url",
```
Comment (pranavrajs) [inline app/javascript/dashboard/i18n/locale/en/integrations.json:106]:
```suggestion
        "URL_LABEL": "Endpoint",
```

File: app/javascript/dashboard/i18n/locale/en/integrations.json [hunk 9]
```diff
@@ -81,6 +84,51 @@
...
+        "URL_PLACEHOLDER": "Enter your app url",
```
Comment (pranavrajs) [inline app/javascript/dashboard/i18n/locale/en/integrations.json:107]:
```suggestion
        "URL_PLACEHOLDER": "Enter the endpoint URL where your app is hosted",
```

File: app/javascript/dashboard/i18n/locale/en/integrations.json [hunk 10]
```diff
@@ -81,6 +84,51 @@
...
+        "URL_ERROR": "Valid URL required"
```
Comment (pranavrajs) [inline app/javascript/dashboard/i18n/locale/en/integrations.json:108]:
```suggestion
        "URL_ERROR": "A valid URL is required"
```

File: app/javascript/dashboard/i18n/locale/en/integrations.json [hunk 11]
```diff
@@ -81,6 +84,51 @@
...
+        "FORM_SUBMIT": "Submit",
+        "FORM_CANCEL": "Cancel",
+        "API_SUCCESS": "Dashboard app configured succesfully",
+        "API_ERROR": "Could not create app, Please try again later"
+      },
+      "UPDATE": {
+        "HEADER": "Edit dashboard app",
+        "FORM_SUBMIT": "Update",
+        "FORM_CANCEL": "Cancel",
+        "API_SUCCESS": "Dashboard app updated succesfully",
```
Comment (pranavrajs) [inline app/javascript/dashboard/i18n/locale/en/integrations.json:121]:
```suggestion
        "API_SUCCESS": "Dashboard app updated successfully",
```

File: app/javascript/dashboard/i18n/locale/en/integrations.json [hunk 12]
```diff
@@ -81,6 +84,51 @@
...
+        "TITLE_ERROR": "App name required",
+        "URL_LABEL": "Iframe url",
+        "URL_PLACEHOLDER": "Enter your app url",
+        "URL_ERROR": "Valid URL required"
+      },
+      "CREATE": {
+        "HEADER": "Add new dashboard app",
+        "FORM_SUBMIT": "Submit",
+        "FORM_CANCEL": "Cancel",
+        "API_SUCCESS": "Dashboard app configured succesfully",
```
Comment (pranavrajs) [inline app/javascript/dashboard/i18n/locale/en/integrations.json:114]:
```suggestion
        "API_SUCCESS": "Dashboard app configured successfully",
```

File: app/javascript/dashboard/i18n/locale/en/integrations.json [hunk 13]
```diff
@@ -81,6 +84,51 @@
...
+        "API_ERROR": "Could not create app, Please try again later"
+      },
+      "UPDATE": {
+        "HEADER": "Edit dashboard app",
+        "FORM_SUBMIT": "Update",
+        "FORM_CANCEL": "Cancel",
+        "API_SUCCESS": "Dashboard app updated succesfully",
+        "API_ERROR": "Could not update app, Please try again later"
```
Comment (pranavrajs) [inline app/javascript/dashboard/i18n/locale/en/integrations.json:122]:
```suggestion
        "API_ERROR": "We couldn't update the app. Please try again later"
```

File: app/javascript/dashboard/i18n/locale/en/integrations.json [hunk 14]
```diff
@@ -81,6 +84,51 @@
...
+        "URL_LABEL": "Iframe url",
+        "URL_PLACEHOLDER": "Enter your app url",
+        "URL_ERROR": "Valid URL required"
+      },
+      "CREATE": {
+        "HEADER": "Add new dashboard app",
+        "FORM_SUBMIT": "Submit",
+        "FORM_CANCEL": "Cancel",
+        "API_SUCCESS": "Dashboard app configured succesfully",
+        "API_ERROR": "Could not create app, Please try again later"
```
Comment (pranavrajs) [inline app/javascript/dashboard/i18n/locale/en/integrations.json:115]:
```suggestion
        "API_ERROR": "We could not create an app. Please try again later"
```

File: app/javascript/dashboard/i18n/locale/en/integrations.json [hunk 15]
```diff
@@ -81,6 +84,51 @@
...
+      },
+      "UPDATE": {
+        "HEADER": "Edit dashboard app",
+        "FORM_SUBMIT": "Update",
+        "FORM_CANCEL": "Cancel",
+        "API_SUCCESS": "Dashboard app updated succesfully",
+        "API_ERROR": "Could not update app, Please try again later"
+      },
+      "DELETE": {
+        "CONFIRM_YES": "Yes Delete %{appName}",
```
Comment (pranavrajs) [inline app/javascript/dashboard/i18n/locale/en/integrations.json:125]:
```suggestion
        "CONFIRM_YES": "Yes, delete it",
```

File: app/javascript/dashboard/i18n/locale/en/integrations.json [hunk 16]
```diff
@@ -81,6 +84,51 @@
...
+        "CONFIRM_NO": "No Keep %{appName}",
```
Comment (pranavrajs) [inline app/javascript/dashboard/i18n/locale/en/integrations.json:126]:
```suggestion
        "CONFIRM_NO": "No, keep it",
```

File: app/javascript/dashboard/i18n/locale/en/integrations.json [hunk 17]
```diff
@@ -81,6 +84,51 @@
...
+        "TITLE": "Confirm Deletion",
```
Comment (pranavrajs) [inline app/javascript/dashboard/i18n/locale/en/integrations.json:127]:
```suggestion
        "TITLE": "Confirm deletion",
```

File: app/javascript/dashboard/i18n/locale/en/integrations.json [hunk 18]
```diff
@@ -81,6 +84,51 @@
...
+        "MESSAGE": "Are you sure to delete %{appName}",
```
Comment (pranavrajs) [inline app/javascript/dashboard/i18n/locale/en/integrations.json:128]:
```suggestion
        "MESSAGE": "Are you sure to delete the app - %{appName}?",
```

File: app/javascript/dashboard/i18n/locale/en/integrations.json [hunk 19]
```diff
@@ -81,6 +84,51 @@
...
+        "API_SUCCESS": "Dashboard app deleted succesfully",
```
Comment (pranavrajs) [inline app/javascript/dashboard/i18n/locale/en/integrations.json:129]:
```suggestion
        "API_SUCCESS": "Dashboard app deleted successfully",
```

File: app/javascript/dashboard/i18n/locale/en/integrations.json [hunk 20]
```diff
@@ -81,6 +84,51 @@
...
+        "API_ERROR": "Could not delete app, Please try again later"
```
Comment (pranavrajs) [inline app/javascript/dashboard/i18n/locale/en/integrations.json:130]:
```suggestion
        "API_ERROR": "We could not delete the app. Please try again later"
```

File: app/javascript/dashboard/i18n/locale/en/integrations.json [hunk 21]
```diff
@@ -81,6 +84,51 @@
     },
     "CONNECT": {
       "BUTTON_TEXT": "Connect"
+    },
+    "DASHBOARD_APPS": {
+      "HEADER_BTN_TXT": "Add new dashboard app",
+      "SIDEBAR_TXT": "<p><b>Dashboard Apps</b> </p> <p>Dashboard apps allow you to create custom pages(apps) right inside your conversations page in chatwoot dashboard. Chatwoot sends the current conversation context(a JSON payload) which contains the user datails and conversation as a window event, which you can use to create your own pages.</p><p>You can read more about dashboard apps here and the demo here in the <a href='#'>docs</a></p>",
```
Comment (pranavrajs) [inline app/javascript/dashboard/i18n/locale/en/integrations.json:90]:
```suggestion
      "SIDEBAR_TXT": "<p><b>Dashboard Apps</b></p><p>Dashboard Apps allow organizations to embed an application inside the Chatwoot dashboard to provide the context for customer support agents. This feature allows you to create an application independently and embed that inside to provide user information, their orders, or their previous payment history.</p><p>When you are embedding your application using the dashboard in Chatwoot, your application will get the context of the conversation and contact as a window event. To receive the context, implement a listener for the message event on your page.</p><p>To add a new dashboard app, you can click on the button "Add a new dashboard app</p>",
```

File: app/javascript/dashboard/i18n/locale/en/integrations.json [hunk 22]
```diff
@@ -81,6 +84,51 @@
...
+      "LIST": {
+        "404": "There are no dashboard apps configured yet",
+        "LOADING": "Fetching dashboard apps",
+        "TABLE_HEADER": [
+          "Name",
+          "Type",
```
Comment (pranavrajs) [inline app/javascript/dashboard/i18n/locale/en/integrations.json:96]: Let us not show the type.

File: app/javascript/dashboard/i18n/locale/en/integrations.json [hunk 23]
```diff
@@ -81,6 +84,51 @@
     },
     "CONNECT": {
       "BUTTON_TEXT": "Connect"
+    },
+    "DASHBOARD_APPS": {
+      "HEADER_BTN_TXT": "Add new dashboard app",
+      "SIDEBAR_TXT": "<p><b>Dashboard Apps</b> </p> <p>Dashboard apps allow you to create custom pages(apps) right inside your conversations page in chatwoot dashboard. Chatwoot sends the current conversation context(a JSON payload) which contains the user datails and conversation as a window event, which you can use to create your own pages.</p><p>You can read more about dashboard apps here and the demo here in the <a href='#'>docs</a></p>",
+      "LIST": {
+        "404": "There are no dashboard apps configured yet",
```
Comment (pranavrajs) [inline app/javascript/dashboard/i18n/locale/en/integrations.json:92]:
```suggestion
        "404": "There are no dashboard apps configured on this account  yet",
```

File: app/javascript/dashboard/routes/dashboard/settings/integrations/Index.vue (hunk 3 above)
Comment (pranavrajs) [reply app/javascript/dashboard/routes/dashboard/settings/integrations/Index.vue:25]: Dashboard Apps allow organizations to embed an application inside the Chatwoot dashboard to provide the context for customer support agents. This feature allows you to create an application independently and embed that inside to provide user information, their orders, or their previous payment history.

**Comment by aswindevps** [comment]
@fayazara What is the expected behavior if the iframe URL endpoint refused to connect?

**Comment by fayazara** [comment]
> @fayazara What is the expected behavior if the iframe URL endpoint refused to connect?

I believe the browser handles it by itself, it will show an error window. "Refused the connect".

**Review by pranavrajs — APPROVED** [review]

**Pull request merged**